dex/networks/zec/tx.go

```diff
@@ -0,0 +1,896 @@
+			return nil, fmt.Errorf("error writing lock_time: %w", err)
+		}
+
+		// nExpiryHeight
+		if err = putUint32(w, tx.ExpiryHeight); err != nil {
+			return nil, fmt.Errorf("error writing nExpiryHeight: %w", err)
+		}
+	}
+
+	// tx_in_count
+	if err = wire.WriteVarInt(w, pver, uint64(len(tx.MsgTx.TxIn))); err != nil {
+		return nil, fmt.Errorf("error writing tx_in_count: %w", err)
+	}
+
+	// tx_in
+	for vin, ti := range tx.TxIn {
+		if err = writeTxIn(w, tx.Version, ti); err != nil {
+			return nil, fmt.Errorf("error writing tx_in %d: %w", vin, err)
+		}
+	}
+
+	// tx_out_count
+	if err = wire.WriteVarInt(w, pver, uint64(len(tx.TxOut))); err != nil {
+		return nil, fmt.Errorf("error writing tx_out_count: %w", err)
+	}
+
+	// tx_out
+	for vout, to := range tx.TxOut {
+		if err = wire.WriteTxOut(w, pver, tx.Version, to); err != nil {
+			return nil, fmt.Errorf("error writing tx_out %d: %w", vout, err)
+		}
+	}
+
+	if tx.Version <= VersionSapling {
+		// lock_time
+		if err = putUint32(w, tx.LockTime); err != nil {
+			return nil, fmt.Errorf("error writing lock_time: %w", err)
+		}
+
+		if tx.Version >= VersionOverwinter {
+			// nExpiryHeight
+			if err = putUint32(w, tx.ExpiryHeight); err != nil {
+				return nil, fmt.Errorf("error writing nExpiryHeight: %w", err)
+			}
+		}
+	}
+
+	if tx.Version == VersionSapling {
+		// valueBalanceSapling
+		if err = putUint64(w, 0); err != nil {
+			return nil, fmt.Errorf("error writing valueBalanceSapling: %w", err)
+		}
+	}
+
+	if tx.Version >= VersionSapling {
+		// nSpendsSapling
+		if err = wire.WriteVarInt(w, pver, 0); err != nil {
+			return nil, fmt.Errorf("error writing nSpendsSapling: %w", err)
+		}
+
+		// nOutputsSapling
+		if err = wire.WriteVarInt(w, pver, 0); err != nil {
+			return nil, fmt.Errorf("error writing nOutputsSapling: %w", err)
+		}
+	}
+
+	if tx.Version >= VersionPreOverwinter && tx.Version <= VersionSapling {
+		// nJoinSplit
+		if err = wire.WriteVarInt(w, pver, 0); err != nil {
+			return nil, fmt.Errorf("error writing nJoinSplit: %w", err)
+		}
+		return w.Bytes(), nil
+	}
+
+	// NU 5
+
+	// no anchorSapling, because nSpendsSapling = 0
+	// no bindingSigSapling or valueBalanceSapling, because nSpendsSapling + nOutputsSapling = 0
+
+	if tx.Version == VersionNU5 {
+		// nActionsOrchard
+		if err = wire.WriteVarInt(w, pver, 0); err != nil {
+			return nil, fmt.Errorf("error writing nActionsOrchard: %w", err)
+		}
+	}
+
+	// vActionsOrchard, flagsOrchard, valueBalanceOrchard, anchorOrchard,
+	// sizeProofsOrchard, proofsOrchard, vSpendAuthSigsOrchard, and
+	// bindingSigOrchard are all empty, because nActionsOrchard = 0.
+
+	return w.Bytes(), nil
+}
+
+// see https://zips.z.cash/protocol/protocol.pdf section 7.1
+func DeserializeTx(b []byte) (*Tx, error) {
+	tx := &Tx{MsgTx: new(wire.MsgTx)}
+	r := bytes.NewReader(b)
+	if err := tx.ZecDecode(r); err != nil {
+		return nil, err
+	}
+
+	remains, _ := io.ReadAll(r)
+	if len(remains) > 0 {
```

Comment:
I believe this is `remains := r.Len() > 0`, so no need for the `ReadAll` to determine this.